src/main/java/ru/javawebinar/topjava/util/UserMealsUtil.java

```diff
@@ -0,0 +1,125 @@
+            }
+            return o1.getDateTime().isBefore(o2.getDateTime()) ? -1 : 1;
+        });
+        //collection for return
+        List<UserMealWithExceed> userMealWithExceedList = new ArrayList<>();
+        List<UserMeal> tempUserMealList = new ArrayList<>();
+        tempUserMealList.add(mealList.get(0));
+        int caloriesCounterOutLoop = mealList.get(0).getCalories();
+        //iterate every except first
+        for (int i = 1; i < mealList.size(); i++) {
+            UserMeal userMeal = mealList.get(i);
+            UserMeal userMealPrevious = mealList.get(i - 1);
+            // add to temp list if the same day and summarize calories
+            if (!userMeal.getDateTime().toLocalDate().isAfter(userMealPrevious.getDateTime().toLocalDate())) {
+                tempUserMealList.add(userMeal);
+                caloriesCounterOutLoop += userMeal.getCalories();
+            } else {
+                //else check all from temp list and add to result list by filter
+                int caloriesCounterInLoop = caloriesCounterOutLoop;
+                tempUserMealList.forEach(userMealFromTemp -> {
+                    LocalTime userMealTime = userMealFromTemp.getDateTime().toLocalTime();
+                    Boolean isExceeded = caloriesCounterInLoop > caloriesPerDay;
+                    if (TimeUtil.isBetween(userMealTime, startTime, endTime)) {
+                        userMealWithExceedList.add(new UserMealWithExceed(userMealFromTemp.getDateTime(),
+                                userMealFromTemp.getDescription(), userMealFromTemp.getCalories(),
+                                isExceeded));
+                    }
+                });
+                caloriesCounterOutLoop = userMeal.getCalories();
+                tempUserMealList.clear();
+                tempUserMealList.add(userMeal);
+            }
+        }
+        return userMealWithExceedList;
+    }
+
+    /**
+     * get filtered using Map without sorting.     *
+     * @param mealList
+     * @param startTime
+     * @param endTime
+     * @param caloriesPerDay
+     * @return
+     *
+     * ----->>>> can this method be considered as completed "Optional (Java 8 Stream API)"???
+     */
+    public static List<UserMealWithExceed> getFilteredWithExceededMap(List<UserMeal> mealList, LocalTime startTime, LocalTime endTime, int caloriesPerDay) {
+        Map<LocalDate, Integer> map = new HashMap<>();
+        // map userMeal to group by date, value is summarized calories of same day
+        mealList.forEach(userMeal -> {
+            LocalDate localDate = userMeal.getDateTime().toLocalDate();
+            map.merge(localDate,userMeal.getCalories(),Integer::sum);
+        });
+        //stream and add to result list by filter
+        List<UserMealWithExceed> mealWithExceeds = mealList.stream()
+                .filter(userMeal -> userMeal.getDateTime().toLocalTime().isAfter(startTime) && userMeal.getDateTime().toLocalTime().isBefore(endTime))
```

Comment:
How to understand that? Please refactor this code. Extract small methods with meaningful names:
```java
   .filter(userMeal -> isStartTimeValid(userMeal) && isEndTimeValid(userMeal))
```